url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/ColliderImpl.java

```diff
@@ -0,0 +1,798 @@
+            }
+
+            m_threadPool.stopAndWait();
+        }
+        catch (final IOException ex)
+        {
+            if (s_logger.isLoggable(Level.WARNING))
+                s_logger.warning( ex.toString() );
+        }
+        catch (final InterruptedException ex)
+        {
+            if (s_logger.isLoggable(Level.WARNING))
+                s_logger.warning( ex.toString() );
+        }
+
+        for (Map.Entry<Integer, RetainableDataBlockCache> me : m_dataBlockCache.entrySet())
+            me.getValue().clear( s_logger );
+        m_dataBlockCache.clear();
+
+        if (s_logger.isLoggable(Level.FINE))
+            s_logger.fine( "finish (" + statLoopIt + ", " + statLoopReadersG0 + ")." );
+    }
+
+    public void stop()
+    {
+        if (s_logger.isLoggable(Level.FINE))
+            s_logger.fine( "" );
+
+        m_lock.lock();
+        try
+        {
+            if (m_stop)
+                return;
+            m_stop = true;
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+
+        executeInThreadPool( new Stopper1() );
+    }
+
+    public final void executeInSelectorThread( SelectorThreadRunnable runnable )
+    {
+        assert( runnable.nextSelectorThreadRunnable == null );
+        final SelectorThreadRunnable tail = s_strTailUpdater.getAndSet( this, runnable );
+        if (tail == null)
+        {
+            m_strHead = runnable;
+
+            for (;;)
+            {
+                SelectorAlarm alarm = m_alarm.get();
+                if (alarm == null)
+                {
+                    m_threadPool.execute( new SelectorAlarm(runnable) );
+                    break;
+                }
+                else if (m_alarm.compareAndSet(alarm, null))
+                {
+                    alarm.cmp = runnable;
+                    m_threadPool.execute( alarm );
+                    break;
+                }
+            }
+        }
+        else
+            tail.nextSelectorThreadRunnable = runnable;
+    }
+
+    public final void executeInSelectorThreadNoWakeup( SelectorThreadRunnable runnable )
+    {
+        assert( runnable.nextSelectorThreadRunnable == null );
+        final SelectorThreadRunnable tail = s_strTailUpdater.getAndSet( this, runnable );
+        if (tail == null)
+            m_strHead = runnable;
+        else
+            tail.nextSelectorThreadRunnable = runnable;
+    }
+
+    public final void executeInSelectorThreadLater( SelectorThreadRunnable runnable )
+    {
+        assert( runnable.nextSelectorThreadRunnable == null );
+        runnable.nextSelectorThreadRunnable = m_strLater;
+        m_strLater = runnable;
+    }
+
+    public final void executeInThreadPool( ThreadPool.Runnable runnable )
+    {
+        m_threadPool.execute( runnable );
+    }
+
+    public void addAcceptor( Acceptor acceptor ) throws IOException
+    {
+        final ServerSocketChannel serverSocketChannel = ServerSocketChannel.open();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *serverSocketChannel*. The resource is referenced at lines 574, 576, 577, 578, 582. The resource is closed at line 614. There are other execution paths that do not contain closure statements, for example, when ServerSocket.bind throws an exception. Ensure *serverSocketChannel* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).